torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -957,6 +957,90 @@ void IndexSwizzle::handle(Expr* e) {
   }
 }
 
+namespace {
+
+// Used for local and shared index mapping
+std::unordered_map<kir::ForLoop*, kir::Val*> indexMapFromTV(
+    const TensorView* tv,
+    const std::vector<kir::ForLoop*>& loops,
+    const std::pair<kir::ForLoop*, int64_t>& alloc_point) {
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  auto alloc_loop = alloc_point.first;
+
+  bool within_alloc = false;
+  if (alloc_loop == nullptr) {
+    within_alloc = true;
+  }
+
+  const auto zero = ir_builder.create<kir::Int>(0);
+
+  const bool is_global = tv->getMemoryType() == MemoryType::Global;
+  const bool is_shared = tv->getMemoryType() == MemoryType::Shared;
+  const bool is_local = tv->getMemoryType() == MemoryType::Local;
+
+  std::unordered_map<kir::ForLoop*, kir::Val*> loop_to_ind_map;
+
+  for (auto loop : loops) {
+    kir::Val* idx = nullptr;
+    // See also LoopNestGenerator::pushAlloc.
+    // NOLINTNEXTLINE(bugprone-branch-clone)
+    if (!within_alloc) {
+      if ((loop->iter_domain()->isThreadDim() && is_shared) ||
+          (loop->iter_domain()->isThread() && is_global)) {
+        idx = loop->index();
+      } else {
+        idx = zero;
+      }
+    } else if (
+        (loop->iter_domain()->isBlockDim() && is_shared) ||
+        (loop->iter_domain()->isThread() && is_local) || loop->vectorize()) {
+      idx = zero;
+    } else {
+      idx = loop->index();
+    }
+
+    loop_to_ind_map[loop] = idx;
+
+    if (!within_alloc && loop == alloc_loop) {
+      within_alloc = true;
+    }
+  }
+  // NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDeleteLeaks)
+  return loop_to_ind_map;
+}
+
+// Map everything we can from reference to provided tv using the provided
+// compute at map. We can't simply try to use the provided tv root domains and
+// map those to the reference as the provided tv may have root domains that
+// don't exist in reference. This can happen when the provided tv is from before
+// a view, but all the loops are generated from TVs generated after the view
+// operation.
+std::unordered_map<IterDomain*, IterDomain*> indexMapReferenceTo(
+    const TensorView* tv,
+    const ComputeAtMap& ca_map,
+    const std::unordered_map<IterDomain*, IterDomain*>&
+        reference_concrete_to_id_map) {
+  std::unordered_map<IterDomain*, IterDomain*> index_map_ref_to_producer;
+  auto p_root_vals = ir_utils::filterByType<Val>(tv->getRootDomain());
+  auto p_domain_vals = ir_utils::filterByType<Val>(tv->domain()->domain());
+  auto all_pid_vals = DependencyCheck::getAllValsBetween(
+      {p_root_vals.begin(), p_root_vals.end()},
+      {p_domain_vals.begin(), p_domain_vals.end()});
```

Comment:
Actually, these create new containers, so no need to have `p_root_vals` and `p_domain_vals`. 

```suggestion
      {tv->getRootDomain().begin(), tv->getRootDomain().end()},
      {tv->domain()->domain().begin(), tv->domain()->domain().end()});
```